qprofiler2/src/org/qcmg/qprofiler2/bam/TagSummaryReport2.java

```diff
@@ -2,433 +2,141 @@
-
-		logger.debug("mapQMatrix.size(): " + mapQMatrix.size());
-
-		Map<Integer, AtomicLong> map = null;
-
-		for (Entry<Integer, MAPQMatrix> entry : mapQMatrix.entrySet()) {
-			logger.debug( entry.getKey() + ", entry.getValue().toString(): " + entry.getValue().toString());
-
-			for (MatrixType type : MatrixType.values()) {
-				map = entry.getValue().getMatrixByType(type);
-
-				for (Entry<Integer, AtomicLong> mapEntry : map.entrySet()) {
-
-					switch (type) {
-					case CM:
-						cmMatrix.put(new MAPQMiniMatrix(entry.getKey(), mapEntry.getKey()), mapEntry.getValue());
-						break;
-					case SM:
-						smMatrix.put(new MAPQMiniMatrix(entry.getKey(), mapEntry.getKey()), mapEntry.getValue());
-						break;
-					case LENGTH:
-						lengthMatrix.put(new MAPQMiniMatrix(entry.getKey(), mapEntry.getKey()), mapEntry.getValue());
-						break;
-					case NH:
-						nhMatrix.put(new MAPQMiniMatrix(entry.getKey(), mapEntry.getKey()), mapEntry.getValue());
-						break;
-					case ZM:
-						zmMatrix.put(new MAPQMiniMatrix(entry.getKey(), mapEntry.getKey()), mapEntry.getValue());
-						break;
-					}
-				}
-			}
-			// removing due to memory limitations
-			mapQMatrix.remove(entry.getKey());
-		}
-	}	
-	private void createMatrix(Element bamReportElement ){
-		
-		Map<MAPQMiniMatrix, AtomicLong> cmMatrix = new TreeMap<MAPQMiniMatrix, AtomicLong>();
-		Map<MAPQMiniMatrix, AtomicLong> smMatrix = new TreeMap<MAPQMiniMatrix, AtomicLong>();
-		Map<MAPQMiniMatrix, AtomicLong> lengthMatrix = new TreeMap<MAPQMiniMatrix, AtomicLong>();
-		Map<MAPQMiniMatrix, AtomicLong> nhMatrix = new TreeMap<MAPQMiniMatrix, AtomicLong>();
-		Map<MAPQMiniMatrix, AtomicLong> zmMatrix = new TreeMap<MAPQMiniMatrix, AtomicLong>();
-
-		generateMAPQSubMaps(cmMatrix, smMatrix, lengthMatrix, nhMatrix, zmMatrix);
-
-		logger.debug("cmMatrix(): " + cmMatrix.size());
-		logger.debug("smMatrix(): " + smMatrix.size());
-		logger.debug("lengthMatrix(): " + lengthMatrix.size());
-		logger.debug("nhMatrix(): " + nhMatrix.size());
-		logger.debug("zmMatrix(): " + zmMatrix.size());
-
-		SummaryReportUtils.lengthMapToXml(bamReportElement, "MAPQMatrixCM",null, cmMatrix, null);
-		SummaryReportUtils.lengthMapToXml(bamReportElement, "MAPQMatrixSM",null, smMatrix,null);
-		SummaryReportUtils.lengthMapToXml(bamReportElement, "MAPQMatrixLength",null, lengthMatrix,null);
-		SummaryReportUtils.lengthMapToXml(bamReportElement, "MAPQMatrixNH",null, nhMatrix,null);
-		SummaryReportUtils.lengthMapToXml(bamReportElement, "MAPQMatrixZM",null, zmMatrix,null);
-
-		zmSmMatrix.toXml(bamReportElement, "ZmSmMatrix");
+	private <T> void outputTag(Element ele, String tag,  Map<T, AtomicLong> tallys) {
+				
+		int size = tallys.size();	
+		ele = XmlUtils.createMetricsNode(ele, "tags:"+tag, size);		
+		AtomicInteger no = new AtomicInteger();		
+		tallys.entrySet().removeIf( e-> no.incrementAndGet() > 100 );
+		boolean percent = (size >= 100)? false : true;
 		
-	}	
-	
- 
-	
+		String[] vs = tag.split(":");
```

Comment:
more efficient to use `TabTokenizer` rather than `split`, and in this instance, it looks like you are only dealing with the first entry in the array, in which case it would be better to use `substring` and `indexOf`
We also have a `Constants` class that contains often used strings and chars (like ":")

Reply:
The performance of split on java8  is close to indexOf, especially it isn't inside a loop. I like the idea of Constants, and try to use it more. 
code update

Reply:
interesting - could you please try doing some benchmarking to see if this is the case for your specific example?
Also, as you are only using one of the entries int he array that you generate, I would be surprised if split is more efficient than indexOf and substring - much fewer objects are created and discarded. But the only way to be sure is to test :)

Reply:
I am also surprised the sentence “split is more efficient than indexOf and substring”. It should be the opposite, shouldn’t it?

Reply:
The Constant.Comma made my code bit mess. I should declare the "seperator" on the top of my class rather than call Constants.COMMA or Constants.COLON everywhere. It easy to cause typo. 
There is no difference between:
public static final String seperator = Constants.COLON_STRING;		
or 
public static final String seperator = ":";
in same Class code

Reply:
you can do a static import which would mean you would not need to include the class name. eg. `Constants.COMMA` becomes `COMMA`